nmrfx-processor/src/main/java/org/nmrfx/processor/datasets/vendor/JCAMPData.java

```diff
@@ -17,987 +17,569 @@
-        return f1coef[iDim];
+    public void setSF(int dim, double value) {
+        sf.put(dim, value);
     }
 
     @Override
-    public String getSymbolicCoefs(int iDim) {
-        return f1coefS[iDim];
+    public void resetSF(int dim) {
+        sf.remove(dim);
     }
 
     @Override
     public String[] getSFNames() {
-        int nDim = getNDim();
-        String[] names = new String[nDim];
-        for (int i = 0; i < nDim; i++) {
-            names[i] = ".OBSERVEFREQUENCY," + (i + 1);
+        String[] names = new String[getNDim()];
+        Arrays.fill(names, "");
+        for (int dim = 0; dim < getNDim(); dim++) {
+            names[dim] = getSFLabel(dim).map(Label::normalized).orElse("");
         }
         return names;
     }
 
-    @Override
-    public String[] getSWNames() {
-        int nDim = getNDim();
-        String[] names = new String[nDim];
-        for (int i = 0; i < nDim; i++) {
-            names[i] = "";
-        }
-        return names;
+    private Optional<Label> getSFLabel(int dim) {
+        //XXX Base freq, or observed freq? should we try to add offset?
+        //Previous implementation was using OBSERVE_FREQUENCY but this is not defined in 2D
+        return Stream.of(_OBSERVE_FREQUENCY, $SFO1, $BF1, $BFREQ, $SF)
+                .filter(label -> block.optional(label, dim).isPresent())
+                .findFirst();
     }
 
     @Override
-    public String[] getLabelNames() {
-        int nDim = getNDim();
-        ArrayList<String> names = new ArrayList<>();
-        for (int i = 0; i < nDim; i++) {
-            String name = getTN(i);
-            if (names.contains(name)) {
-                name = name + "_" + (i + 1);
-            }
-            names.add(name);
-        }
-        return names.toArray(new String[names.size()]);
+    public double getSW(int dim) {
+        return sw.computeIfAbsent(dim, this::extractSW);
     }
 
-    @Override
-    public double getSF(int iDim) {
-        return Sf[iDim];
+    private double extractSW(int dim) {
+        return block.optional($SW_H, dim)
+                .map(JCampRecord::getDouble)
+                .orElseThrow(() -> new IllegalStateException("Unknown spectral width, $SW_H undefined for dimension " + dim));
     }
 
     @Override
-    public void setSF(int iDim, double value) {
-        Sf[iDim] = value;
+    public void setSW(int dim, double value) {
+        sw.put(dim, value);
     }
 
     @Override
-    public void resetSF(int iDim) {
+    public void resetSW(int dim) {
+       sw.remove(dim);
     }
 
     @Override
-    public double getSW(int iDim) {
-        return Sw[iDim];
+    public String[] getSWNames() {
+        //XXX check whether this is useful
+        // original JCAMPData returned an array of empty strings
+        String[] names = new String[getNDim()];
+        Arrays.fill(names, "");
+        names[0] = block.optional($SW_H).map(JCampRecord::getNormalizedLabel).orElse("");
+        return names;
```

Comment:
This was not implemented in the original. I can see the change in the UI, I'm not sure how useful it is.

Reply:
This is used in setting up processing scripts so that rather than setting an explcit value for the sweep width (or spectrometer frequency) one can specify the name of the parameter to use.  That way the same script can be used with different conditions.  Some pulse programs (I remember doing this on Varian) used different parameters for a given dimension.  Some times the first indirect dimension might have used "sw1" and other times "sw".  Probably not necessary to support here.